processing/src/main/java/org/apache/druid/query/ChainedExecutionQueryRunner.java

```diff
@@ -141,33 +144,34 @@ public ChainedExecutionQueryRunner(
                           );
                         }
                     )
-                )
-            );
+                );
 
-            queryWatcher.registerQueryFuture(query, futures);
+            ListenableFuture<List<Iterable<T>>> future = Futures.allAsList(futures);
+            queryWatcher.registerQueryFuture(query, future);
 
             try {
               return new MergeIterable<>(
                   ordering.nullsFirst(),
                   QueryContexts.hasTimeout(query) ?
-                      futures.get(QueryContexts.getTimeout(query), TimeUnit.MILLISECONDS) :
-                      futures.get()
+                      future.get(QueryContexts.getTimeout(query), TimeUnit.MILLISECONDS) :
+                      future.get()
               ).iterator();
             }
             catch (InterruptedException e) {
               log.noStackTrace().warn(e, "Query interrupted, cancelling pending results, query id [%s]", query.getId());
-              futures.cancel(true);
+              GuavaUtils.cancelAll(true, ImmutableList.<Future>builder().add(future).addAll(futures).build());
```

Comment:
It seems easy to forget canceling `future` and so error-prone. How about modifying `GuavaUtils.cancelAll()` to take `future` as well? So it would be like

```java
  public static <F extends Future<?>> void cancelAll(
      boolean mayInterruptIfRunning,
      @Nullable ListenableFuture<?> combinedFuture,
      List<F> futures
  )
  {
    final List<Future> allFuturesToCancel = new ArrayList<>(futures);
    allFuturesToCancel.add(combinedFuture);
    if (allFuturesToCancel.isEmpty()) {
      return;
    }
    allFuturesToCancel.forEach(f -> {
      try {
        f.cancel(mayInterruptIfRunning);
      }
      catch (Throwable t) {
        log.warn(t, "Error while cancelling future.");
      }
    });
  }
```

Reply:
Or, more structured way to do could be adding a new `CombinedFuture` like this

```java
  public static class CombinedFuture<V> implements Future<List<V>>
  {
    private final List<ListenableFuture<V>> underlyingFutures;
    private final ListenableFuture<List<V>> combined;
    
    public CombinedFuture(List<ListenableFuture<V>> futures)
    {
      this.underlyingFutures = futures;
      this.combined = Futures.allAsList(futures);
    }

    @Override
    public boolean cancel(boolean mayInterruptIfRunning)
    {
      if (combined.isDone() || combined.isCancelled()) {
        return false;
      } else {
        cancelAll(mayInterruptIfRunning, combined, underlyingFutures);
        return true;
      }
    }

    @Override
    public boolean isCancelled()
    {
      return combined.isCancelled();
    }

    @Override
    public boolean isDone()
    {
      return combined.isDone();
    }

    @Override
    public List<V> get() throws InterruptedException, ExecutionException
    {
      return combined.get();
    }

    @Override
    public List<V> get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException
    {
      return combined.get(timeout, unit);
    }
  }
```

I'm fine with either way.

Reply:
> It seems easy to forget canceling `future` and so error-prone. How about modifying `GuavaUtils.cancelAll()` to take `future` as well? So it would be like
> 
> ```java
>   public static <F extends Future<?>> void cancelAll(
>       boolean mayInterruptIfRunning,
>       @Nullable ListenableFuture<?> combinedFuture,
>       List<F> futures
>   )
>   {
>     final List<Future> allFuturesToCancel = new ArrayList<>(futures);
>     allFuturesToCancel.add(combinedFuture);
>     if (allFuturesToCancel.isEmpty()) {
>       return;
>     }
>     allFuturesToCancel.forEach(f -> {
>       try {
>         f.cancel(mayInterruptIfRunning);
>       }
>       catch (Throwable t) {
>         log.warn(t, "Error while cancelling future.");
>       }
>     });
>   }
> ```

Well, thanks for your tips, i'll follow it except one point. It's better to cancel the `combinedFuture` first, because if we cancel the first future in  `underlyingFutures`  , it will trigger the listener of `com.google.common.util.concurrent.Futures.CombinedFuture` which is added for every future in `underlyingFutures`  by `init` method. This listener is actually a method called `setOneValue`  which will set combinedFuture's status as `CANCELLED` rather than `INTERRUTED` as we expect. In addition , the listener of `combinedFuture` will set the status of other future in `underlyingFutures`   as the same with itself(`CANCELLED` rather than `INTERRUTED` as we expect). I have test it in the test of `testQueryTimeout` in `ChainedExecutionQueryRunnerTest`.

Reply:
Ah makes sense 👍. Would you please add a comment on [this line](https://github.com/apache/druid/pull/10027/files/7fc9d5036af8003f1a576ccab128b57571ad03fd..a51d7e16f5f15af33fa239419aaa82bffb99f8f5#diff-298d22d65226e0f32251f009130a0032R102) about why we cancel it first? It would help other people to understand your intention. Maybe the comment can say "Canceling combinedFuture first so that it can complete with `INTERRUPTED` as its final state. See ChainedExecutionQueryRunnerTest.testQueryTimeout()".